[World Leaks] Investigate leaks in LayoutTests/imported/w3c/web-platform-tests/webaudio/the-audio-api/the-audioworklet-interface/

nipunshukla21 opened the pull request:
#### eadd64122b83781dc378218c5a073f9935bec2b9
<pre>
[World Leaks] Investigate leaks in LayoutTests/imported/w3c/web-platform-tests/webaudio/the-audio-api/the-audioworklet-interface/
<a href="https://bugs.webkit.org/show_bug.cgi?id=299312">https://bugs.webkit.org/show_bug.cgi?id=299312</a>
<a href="https://rdar.apple.com/161118548">rdar://161118548</a>

Reviewed by Chris Dumez.

AudioWorkletMessagingProxy holds a strong const Ref&lt;Document&gt;. Removing this reference and
using a ScriptExecutionContextIdentifier to post tasks when necessary fixes approximately
30 leaky tests.

Test: webaudio/audioworklet-does-not-leak.html

* LayoutTests/webaudio/audioworklet-does-not-leak-expected.txt: Added.
* LayoutTests/webaudio/audioworklet-does-not-leak.html: Added.
* LayoutTests/webaudio/resources/audioworklet-leak-test-frame.html: Added.
* Source/WebCore/Modules/webaudio/AudioWorkletMessagingProxy.cpp:
(WebCore::AudioWorkletMessagingProxy::AudioWorkletMessagingProxy):
(WebCore::AudioWorkletMessagingProxy::createRTCDataChannelRemoteHandlerConnection):
(WebCore::AudioWorkletMessagingProxy::loaderContextIdentifier const):
(WebCore::AudioWorkletMessagingProxy::postTaskToLoader):
(WebCore::AudioWorkletMessagingProxy::postTaskToAudioWorklet):
* Source/WebCore/Modules/webaudio/AudioWorkletMessagingProxy.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::messagePortDisentangled):

Canonical link: <a href="https://commits.webkit.org/301328@main">https://commits.webkit.org/301328@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c2c4ccbaff5f73064c3d0b2337242dd0f4036917

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125603 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45265 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36012 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132465 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/77531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127474 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45952 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53825 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95682 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63798 "Unable to confirm if test failures are introduced by change, retrying build") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128551 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36726 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112312 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76179 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35635 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30494 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75938 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106512 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30712 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135137 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52403 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40153 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104153 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52844 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108528 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103885 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26458 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49241 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27544 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/49614 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52293 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58090 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51645 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54997 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53340 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/webaudio/AudioWorkletMessagingProxy.cpp:
```diff
@@ -97,24 +97,24 @@ RefPtr<CacheStorageConnection> AudioWorkletMessagingProxy::createCacheStorageCon
 {
     ASSERT(isMainThread());
-    if (!m_document->page())
+    if (!m_document.get().page())
```
All these `.get()` changes shouldn't be needed, you can totally use `->` with a WeakRef or a WeakPtr or a CheckedRef or a CheckedPtr. Please revert them.

cdumez commented on Source/WebCore/Modules/webaudio/AudioWorkletMessagingProxy.h:
```diff
@@ -73,7 +74,7 @@ class AudioWorkletMessagingProxy final : public WorkletGlobalScopeProxy, public
 
     WeakPtr<AudioWorklet> m_worklet;
-    const Ref<Document> m_document;
+    const WeakRef<Document, WeakPtrImplWithEventTargetData> m_document;
```
This looks reasonable, however, based on the test failure, we should likely use a WeakPtr and do null checks.

Also please take a look at Safer CPP failures. Odds are the document now needs to be ref'd on the stack before being used.

nipunshukla21 replied:
This is related to why I was using `WeakRef::get()` as using `WeakRef::operator->()` asserts on `canSafelyBeUsed()`, while `get()` doesn't. Something similar happens when I use a WeakPtr member variable and then create a strong reference on the stack -- leading to these debug assert crashes in tests:
```
20:49:04.697 71491   ASSERTION FAILED: canSafelyBeUsed()
20:49:04.697 71491   /Users/nipunshukla/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/WeakPtr.h(97) : T *WTF::WeakPtr<WebCore::Document, WebCore::WeakPtrImplWithEventTargetData>::get() const [T = WebCore::Document, WeakPtrImpl = WebCore::WeakPtrImplWithEventTargetData, PtrTraits = WTF::RawPtrTraits<WebCore::WeakPtrImplWithEventTargetData>]
20:49:04.697 71491   1   0x300653b48 WTF::WeakPtr<WebCore::Document, WebCore::WeakPtrImplWithEventTargetData, WTF::RawPtrTraits<WebCore::WeakPtrImplWithEventTargetData>>::get() const
20:49:04.697 71491   2   0x3041f778c WebCore::AudioWorkletMessagingProxy::postTaskToLoader(WebCore::ScriptExecutionContext::Task&&)
```
I believe what's happening is that we create the WeakPtr member variable in the ctor on the main thread, but then use it and hit the assert in functions like `postTaskToLoader()` which are called off the main thread when we either use `RefPtr document = m_document.get()` or `WeakPtr::operator->()`-- causing us to fail [this assert](https://searchfox.org/wubkat/rev/227c6a7e9cad3099da4c496dbc311de75808f9b8/Source/WTF/wtf/WeakPtr.h#168). I also can't do something like `RefPtr document { m_document }` as that fails the type checker. Do you know of a good way to get around this? For now I've changed the code to use `callOnMainThread()` and variants where necessary, which seem to pass tests without asserting or crashing, but I'm not very happy with it.

nipunshukla21 commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1626,8 +1626,7 @@ void NetworkConnectionToWebProcess::entangleLocalPortInThisProcessToRemote(const
 void NetworkConnectionToWebProcess::messagePortDisentangled(const MessagePortIdentifier& port)
 {
-    auto result = m_processEntangledPorts.remove(port);
-    ASSERT_UNUSED(result, result);
```
Removing this assert seems to fix the flakiness and I think it's safe due to didDisentangleMessagePort() being a no-op if `m_channels.get(port)` doesn't find anything.

cdumez replied:
Please make sure this change is still needed after integrating my proposals above. It'd be unfortunate to drop an assertion if we could avoid it.

cdumez commented on Source/WebCore/Modules/webaudio/AudioWorkletMessagingProxy.cpp:
```diff
@@ -97,26 +97,39 @@ RefPtr<CacheStorageConnection> AudioWorkletMessagingProxy::createCacheStorageCon
 ScriptExecutionContextIdentifier AudioWorkletMessagingProxy::loaderContextIdentifier() const
 {
-    return m_document->identifier();
+    ScriptExecutionContextIdentifier identifier = WTF::MarkableTraits<ScriptExecutionContextIdentifier>::emptyValue();
```
We should just use `std::optional< ScriptExecutionContextIdentifier>` here instead of assigning the empty value, which violates the fact that our strong identifiers have to be valid.

cdumez commented on Source/WebCore/Modules/webaudio/AudioWorkletMessagingProxy.cpp:
```diff
@@ -97,26 +97,39 @@ RefPtr<CacheStorageConnection> AudioWorkletMessagingProxy::createCacheStorageCon
-    return m_document->identifier();
+    ScriptExecutionContextIdentifier identifier = WTF::MarkableTraits<ScriptExecutionContextIdentifier>::emptyValue();
+    callOnMainThreadAndWait([weakDocument = m_document, &identifier] {
+        if (RefPtr document = weakDocument.get())
```
I don't think you need to ref the document here? `identifier()` is a trivial inline function.

cdumez commented on Source/WebCore/Modules/webaudio/AudioWorkletMessagingProxy.cpp:
```diff
@@ -97,26 +97,39 @@ RefPtr<CacheStorageConnection> AudioWorkletMessagingProxy::createCacheStorageCon
+        if (RefPtr document = weakDocument.get())
+            identifier = document->identifier();
+    });
+    return identifier;
```
Returning an "empty" identifier here would be a bad bug as our codebase assumes those identifiers are always valid.

cdumez replied:
My proposal for this function:
1. Add a `ScriptExecutionContextIdentifier m_documentIdentifier` data member.
2. Initialize this data member in the class destructor (which I believe runs on the main thread)
3. Just `return m_documentIdentifier;` in this function without worrying about threading.

cdumez commented on Source/WebCore/Modules/webaudio/AudioWorkletMessagingProxy.cpp:
```diff
@@ -97,26 +97,39 @@ RefPtr<CacheStorageConnection> AudioWorkletMessagingProxy::createCacheStorageCon
 void AudioWorkletMessagingProxy::postTaskToLoader(ScriptExecutionContext::Task&& task)
 {
-    m_document->postTask(WTFMove(task));
+    callOnMainThread([weakDocument = m_document, task = WTFMove(task)]() mutable {
```
If you follow my advice above and store the document identifier as data member then you no longer need the document pointer to post a task. You can use `ScriptExecutionContext::postTaskTo(ScriptExecutionContextIdentifier)`.

cdumez commented on Source/WebCore/Modules/webaudio/AudioWorkletMessagingProxy.cpp:
```diff
@@ -97,26 +97,39 @@ RefPtr<CacheStorageConnection> AudioWorkletMessagingProxy::createCacheStorageCon
-    m_document->postTask([protectedThis = Ref { *this }, task = WTFMove(task)](ScriptExecutionContext&) {
-        if (protectedThis->m_worklet)
-            task(*protectedThis->m_worklet);
+    callOnMainThread([weakDocument = m_document, protectedThis = Ref { *this }, task = WTFMove(task)]() mutable {
```
Ditto.

cdumez commented on Source/WebCore/Modules/webaudio/AudioWorkletMessagingProxy.h:
```diff
@@ -73,7 +74,7 @@ class AudioWorkletMessagingProxy final : public WorkletGlobalScopeProxy, public
 
     WeakPtr<AudioWorklet> m_worklet;
-    const Ref<Document> m_document;
+    const WeakPtr<Document, WeakPtrImplWithEventTargetData> m_document;
```
Curious if you still need this data member after my proposal. We should check.

cdumez reviewed (approved):
Looks much better! r+

nipunshukla21 committed:
[World Leaks] Investigate leaks in LayoutTests/imported/w3c/web-platform-tests/webaudio/the-audio-api/the-audioworklet-interface/
https://bugs.webkit.org/show_bug.cgi?id=299312
rdar://161118548

Reviewed by Chris Dumez.

AudioWorkletMessagingProxy holds a strong const Ref<Document>. Removing this reference and
using a ScriptExecutionContextIdentifier to post tasks when necessary fixes approximately
30 leaky tests.

Test: webaudio/audioworklet-does-not-leak.html

* LayoutTests/webaudio/audioworklet-does-not-leak-expected.txt: Added.
* LayoutTests/webaudio/audioworklet-does-not-leak.html: Added.
* LayoutTests/webaudio/resources/audioworklet-leak-test-frame.html: Added.
* Source/WebCore/Modules/webaudio/AudioWorkletMessagingProxy.cpp:
(WebCore::AudioWorkletMessagingProxy::AudioWorkletMessagingProxy):
(WebCore::AudioWorkletMessagingProxy::createRTCDataChannelRemoteHandlerConnection):
(WebCore::AudioWorkletMessagingProxy::loaderContextIdentifier const):
(WebCore::AudioWorkletMessagingProxy::postTaskToLoader):
(WebCore::AudioWorkletMessagingProxy::postTaskToAudioWorklet):
* Source/WebCore/Modules/webaudio/AudioWorkletMessagingProxy.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::messagePortDisentangled):

Canonical link: https://commits.webkit.org/301328@main
